libparsec/crates/platform_mountpoint/src/windows/mod.rs

```diff
@@ -0,0 +1,480 @@
+            Context::Info(info) => info.read(buffer, offset as usize),
+            Context::Dir(_) => unreachable!(),
+        })
+    }
+
+    fn write(
+        &self,
+        file_context: Self::FileContext,
+        buffer: &[u8],
+        offset: u64,
+        mode: WinWriteMode,
+    ) -> Result<usize, NTSTATUS> {
+        let fc = file_context.lock().expect("Mutex is poisoned");
+
+        let mode = WriteMode::from(mode);
+
+        if let Context::File(file) = fc.deref() {
+            Ok(self.fd_write(file.fd, buffer, offset, mode)?)
+        } else {
+            Err(STATUS_NOT_IMPLEMENTED)
+        }
+    }
+
+    fn flush(&self, file_context: Self::FileContext) -> Result<(), NTSTATUS> {
+        let fc = file_context.lock().expect("Mutex is poisoned");
+
+        if let Context::File(file) = fc.deref() {
+            self.fd_flush(file.fd);
+        }
+
+        Ok(())
+    }
+
+    fn set_basic_info(
+        &self,
+        _file_context: Self::FileContext,
+        _file_attributes: FileAttributes,
+        _creation_time: u64,
+        _last_access_time: u64,
+        _last_write_time: u64,
+        _change_time: u64,
+    ) -> Result<(), NTSTATUS> {
+        Err(STATUS_NOT_IMPLEMENTED)
+    }
+
+    fn set_file_size(
+        &self,
+        file_context: Self::FileContext,
+        new_size: u64,
+        set_allocation_size: bool,
+    ) -> Result<(), NTSTATUS> {
+        let fc = file_context.lock().expect("Mutex is poisoned");
+
+        if let Context::File(file) = fc.deref() {
+            self.fd_resize(file.fd, new_size, set_allocation_size)?;
+        }
+
+        Ok(())
+    }
+
+    fn rename(
+        &self,
+        _file_context: Self::FileContext,
+        file_name: &U16CStr,
+        new_file_name: &U16CStr,
+        replace_if_exists: bool,
+    ) -> Result<(), NTSTATUS> {
+        let path = PathBuf::from(file_name.to_os_string());
+        let new_path = PathBuf::from(new_file_name.to_os_string());
+        self.entry_rename(&path, &new_path, replace_if_exists)?;
+        Ok(())
+    }
+
+    fn get_security(
+        &self,
+        _file_context: Self::FileContext,
+    ) -> Result<PSecurityDescriptor, NTSTATUS> {
+        Ok(SD.as_ptr())
+    }
+
+    fn set_delete(
+        &self,
+        _file_context: Self::FileContext,
+        file_name: &U16CStr,
+        _delete_file: bool,
+    ) -> Result<(), NTSTATUS> {
+        let path = PathBuf::from(file_name.to_os_string());
+        let stat = self.entry_info(&path)?;
+
+        if path == Path::new("/") {
+            return Err(STATUS_RESOURCEMANAGER_READ_ONLY);
+        } else if !stat.children.is_empty() {
+            return Err(STATUS_DIRECTORY_NOT_EMPTY);
+        }
+
+        Ok(())
+    }
+}
+
+pub fn mount(mountpoint: &Path) -> FileSystem<MountpointManager> {
+    let mountpoint =
+        U16CString::from_os_str(mountpoint.as_os_str()).expect("Unreachable, valid OsStr");
```

Comment:
what if `mountpoint` contains utf8 characters ? utf8 and utf16 are not equivalent and hence require conversion between the two (which are not done here)

For instance, considering `0xFF` is not valid in utf8:
```python
>>> b"\xff".decode("utf8")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 0: invalid start byte
>>> b"\x00\xff".decode("utf16")
'\uff00'
>>> b"\x00\xff".decode("utf8")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 1: invalid start byte
```

So in this example we can see that the buffer `0x00 0xff` will work with `U16CString::from_os_str` but will cause an error with `U8CString::from_os_str`

So I guess you will have to do conversion between utf8 and utf16 here (but fortunatly it seems [Rust has you covered here ^^](https://doc.rust-lang.org/std/primitive.str.html#method.encode_utf16) )

Reply:
What ? o_O

Reply:
`mountpoint.as_os_str()` doesn't do any conversion (i.e. it only expose the underlying bytes buffer)

Similarly `U16CString::from_os_str` doesn't do any conversion (i.e. it takes the bytes buffer and use it as-is)

Also see  https://docs.rs/widestring/latest/x86_64-pc-windows-msvc/widestring/ustring/struct.U16String.html#method.from_os_str

> Since [OsStr](https://doc.rust-lang.org/nightly/std/ffi/os_str/struct.OsStr.html) makes no guarantees that it is valid data, there is no guarantee that the resulting [U16String](https://docs.rs/widestring/latest/x86_64-pc-windows-msvc/widestring/ustring/struct.U16String.html) will be valid UTF-16.

So what you are doing is equivalent to:
```c
char *my_uft8_string = "'I \xe2\x9d\xa4\xef\xb8\x8f Rust"; // UTF8 of "I ❤️ Rust"
char *my_utf16_string = my_utf8_string;
```
In this example it is trivial to understand `my_utf8_string` and `my_utf16_string` point to the exact same data, which cannot be both utf8 and utf16 !
Here, the "I love Rust" in UTF16 is `\xff\xfeI\x00 \x00d'\x0f\xfe \x00R\x00u\x00s\x00t\x00`, hence doing `my_utf16_string = my_utf8_string` is wrong because then `my_utf16_string` would not contains valid UTF16 data.
Instead we should do something like `my_utf16_string = utf8_to_utf16(my_utf8_string)`.

Reply:
I think I'm wrong here: I was under the assumption that `Path` was always UTF8, but this is wrong: `Path` is a wrapper over `OSStr` and hence has an encoding which is dependent of the platform (hence UTF16 on Windows)

So doing `U16CString::from_os_str(mountpoint.as_os_str())` is always ok given the `OSStr` has been obtained from a `Path`

(on top of that `Path` has to deal with FFI so it cannot guarantee that it contains valid UTF16 data)

so sorry for the noise, everything is fine here ^^